### PR TITLE
[FLINK-16005][yarn] Support yarn and hadoop config override

### PR DESCRIPTION
**Pull request by yuruguo**
## What is the purpose of the change

This pull request allow users to overwrite hadoop (keys in [core-default.xml](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/core-default.xml) and [hdfs-default.xml](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/hdfs-default.xml)) / yarn (keys in [yarn-default.xml](https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-common/yarn-default.xml)) configuration by corresponding set flink configuration started with prefix "flink.hadoop." / "flink.yarn.". It is convenient for users to set hadoop / yarn configuration, because they can only focus on flink configuration.

## Brief change log

  - *Add all configuration key with prefix `flink.hadoop.` in flink conf to hadoop conf*
  - *Add all configuration key with prefix `flink.yarn.` in flink conf to yarn conf*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: ( no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by yuruguo** [comment]
@tillrohrmann Could you review code for this PR, thx!

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/util/HadoopUtils.java [hunk 1]
```diff
@@ -114,6 +117,24 @@ public static Configuration getHadoopConfiguration(
...
+                    result.set(newKey, value);
+                    LOG.debug(
+                            "Adding Flink config entry for {} as {}={} to Hadoop config",
+                            key,
+                            newKey,
+                            value);
+                    foundHadoopConfiguration = true;
+                }
+            }
+        }
```
Comment (tillrohrmann) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/util/HadoopUtils.java:136]: This code block and the one for the Yarn configuration look pretty much the same. Maybe we can deduplicate them.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java [hunk 2]
```diff
@@ -77,6 +78,8 @@ public ApplicationId getClusterId(Configuration configuration) {
     private YarnClusterDescriptor getClusterDescriptor(Configuration configuration) {
         final YarnClient yarnClient = YarnClient.createYarnClient();
         final YarnConfiguration yarnConfiguration = new YarnConfiguration();
+        yarnConfiguration.addResource(HadoopUtils.getHadoopConfiguration(configuration));
+        yarnConfiguration.addResource(Utils.getYarnConfiguration(configuration));
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java:82]: Would it make sense to load the `YarnConfiguration` through a utility. Maybe even `HadoopUtils`? That way it would always be consistent.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java [hunk 3]
```diff
@@ -77,6 +78,8 @@ public ApplicationId getClusterId(Configuration configuration) {
     private YarnClusterDescriptor getClusterDescriptor(Configuration configuration) {
         final YarnClient yarnClient = YarnClient.createYarnClient();
         final YarnConfiguration yarnConfiguration = new YarnConfiguration();
+        yarnConfiguration.addResource(HadoopUtils.getHadoopConfiguration(configuration));
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java:81]: Why is it necessary to add the `HadoopConfiguration` here?

File: docs/_includes/generated/yarn_config_configuration.html [hunk 4]
```diff
@@ -14,6 +14,20 @@
...
+            </td>
+        </tr>
+        <tr>
+            <td><h5>flink.yarn.&lt;key&gt;</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>String</td>
+            <td>A general option to probe Yarn configuration through prefix `flink.yarn.`. Flink will remove the prefix `flink.` to get yarn_key (from <a href="https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-common/yarn-default.xml">yarn-default.xml</a>) then set the yarn_key and value to Yarn configuration. For example, flink.yarn.resourcemanager.container.liveness-monitor.interval-ms=300000 in Flink configuration and convert to yarn.resourcemanager.container.liveness-monitor.interval-ms=300000 in Yarn configuration.
+            </td>
+        </tr>
```
Comment (tillrohrmann) [inline docs/_includes/generated/yarn_config_configuration.html:30]: This file is generated. I think we can add a `ConfigOption` with the key `flink.yarn.<key>` and `flink.hadoop.<key>` and the corresponding descriptions to `YarnConfigOptions`.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks a lot for creating this PR @yuruguo. I think the change looks already quite good. I had a couple of suggestions on minor improvements. Please take a look.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java (hunk 3 above)
Comment (yuruguo) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java:81]: I find this `yarnConfiguration` can be used to get FileSystem in YarnClusterDescriptor.class, such as in`startAppMaster` method:
image
So add `HadoopConfiguration` to benefit the filesystem.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java (hunk 2 above)
Comment (yuruguo) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java:82]: If `HadoopUtils` provides both `getHadoopConfiguration` and `getYarnConfiguration`, then it could be consistent and simplify code https://github.com/apache/flink/pull/14477#discussion_r553198430.
But I am not sure if it is really appropriate and standardized to put `getYarnConfiguration` in the `flink-hadoop-fs` module, because the `flink-yarn` module itself has `Utils`, it may be more clear to use it directly.

File: docs/_includes/generated/yarn_config_configuration.html (hunk 4 above)
Comment (yuruguo) [reply docs/_includes/generated/yarn_config_configuration.html:30]: Thanks for your guidance, I understand its generation mechanism and I will add it.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java:81]: ok, thanks for the clarification.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClientFactory.java:82]: Yes, it is better to put it into the `Utils` class of `flink-yarn`.

File: docs/_includes/generated/yarn_config_configuration.html (hunk 4 above)
Comment (yuruguo) [reply docs/_includes/generated/yarn_config_configuration.html]: @tillrohrmann I have added two options `flink.yarn.<key>` and `flink.hadoop.<key>` in `YarnConfigOptions` and auto generated related description in HTML document.
Then please take a look at this modification and evaluate the completion of this pr, thx!

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java [hunk 5]
```diff
@@ -124,6 +125,8 @@ public YarnResourceManagerDriver(
         super(flinkConfig, GlobalConfiguration.loadConfiguration(configuration.getCurrentDir()));
 
         this.yarnConfig = new YarnConfiguration();
+        yarnConfig.addResource(HadoopUtils.getHadoopConfiguration(flinkConfig));
+        yarnConfig.addResource(Utils.getYarnConfiguration(flinkConfig));
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java:129]: I think we should still put this logic in the `Utils` class and then reuse it.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for the update @yuruguo. I think we can still improve it a bit by moving the logic responsible for loading the `YarnConfiguration` into the `Utils` class and reuse everywhere it is needed.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java (hunk 5 above)
Comment (yuruguo) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java:129]: Ok @tillrohrmann, We can slightly modify the `getYarnConfiguration` method in the `Utils`  to call `HadoopUtils.getHadoopConfiguration(flinkConfig)`, as follows:
image
then `getYarnConfiguration` can be reused everywhere needed,as follows:
in `YarnClusterClientFactory` class
image
in `YarnResourceManagerDriver` class
image

Or another way is to keep the `getYarnConfiguration` method unchanged and add a method `loadYarnConfiguration` packing above logic in the `Utils` class, as follows：
image
then `loadYarnConfiguration` can be reused everywhere needed.

Which one do you think is appropriate? THX!

Comment (tillrohrmann) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java:129]: I would be in favour of only having a single method at them moment. Once we need to load a `YarnConfiguration` w/o the `HadoopConfiguration`, we can split the method. But until then, I would like to keep it simple.

File: flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java [hunk 6]
```diff
@@ -334,18 +334,32 @@
...
-                                    + "and <a href=\"https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/hdfs-default.xml\">hdfs-default.xml</a>) then set the <key> and value to Hadoop configuration. "
-                                    + "For example, flink.hadoop.dfs.replication=5 in Flink configuration and convert to dfs.replication=5 in Hadoop configuration.");
+                            Description.builder()
+                                    .text(
+                                            "A general option to probe Hadoop configuration through prefix 'flink.hadoop.'. Flink will remove the prefix to get <key> (from %s and %s) then set the <key> and value to Hadoop configuration."
+                                                    + " For example, flink.hadoop.dfs.replication=5 in Flink configuration and convert to dfs.replication=5 in Hadoop configuration.",
+                                            link(
+                                                    "https://hadoop.apache.org/docs/r2.4.1/hadoop-project-dist/hadoop-common/core-default.xml",
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:342]: Why did you change the link to Hadoop version 2.4.1 from stable?

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @yuruguo. The changes look good to me. I had one remaining question about the link to the Hadoop configurations. Why did you choose to link v2.4.1 instead of the latest stable version? I think Flink should not been limited to v2.4.1.

File: flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java (hunk 6 above)
Comment (yuruguo) [reply flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:342]: Thanks for your review again @tillrohrmann. 
Because I found that configuration option `yarn.application-attempts` is based on v2.4.1, so I continue to use 2.4.1 for alignment. If unbundling with a specific version is a better way, then I will use the most recent stable version instead, including `yarn.application-attempts`. 
If it's ok, I will update again.

Comment (tillrohrmann) [reply flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:342]: Yes, please update to link to the stable configuration.

**Comment by yuruguo** [comment]
Hi @tillrohrmann, CI has failed but it seems to be unrelated to this pr, please take a look.

**Comment by tillrohrmann** [comment]
The test failures seem indeed to be unrelated. Merging this PR now.
